src/basic/log.c

```diff
@@ -1477,3 +1523,72 @@ void log_setup(void) {
         if (log_on_console() && show_color < 0)
                 log_show_color(true);
 }
+
+LogContext* log_context_attach(LogContext *c) {
+        assert(c);
+
+        STRV_FOREACH(s, c->fields)
+                _log_context_num_fields++;
+
+        return LIST_PREPEND(ll, _log_context, c);
+}
+
+LogContext* log_context_detach(LogContext *c) {
+        if (!c)
+                return NULL;
+
+        STRV_FOREACH(s, c->fields)
+                _log_context_num_fields--;
+
+        LIST_REMOVE(ll, _log_context, c);
+        return NULL;
+}
+
+LogContext* log_context_new(char **fields, bool owned) {
+        if (strv_isempty(fields))
+                return NULL;
```

Comment:
hmm, this makes it hard to distinguish allocation failures from invocations with an empty list.

maybe add:

```c
#define LOG_CONTEXT_EMPTY ((LogContext*) (uintptr_t) 1)
```

as a special value you can return here to indicate an empty log context.

And then handle that specially in log_context_attach(), log_context_free() and so on.

Reply:
or, alternatively, just define a static empty one as:

```c
static const LogContext empty_log_context = {};
```

and then return a pointer to that?

Reply:
I just removed the empty `fields` check, if there are no fields, we still allocate a log context object. We can't use a static const instance here because we always have to return an allocated object that the user can free.